src/deploy/functions/params.ts

```diff
@@ -0,0 +1,342 @@
+  } else if (param.type === "int") {
+    return typeof value === "number" && Number.isInteger(value);
+  } else {
+    assertExhaustive(param);
+  }
+}
+
+/**
+ * A param defined by the SDK may resolve to:
+ * - the value of a Cloud Secret in the same project with name == param name (not implemented yet), but only if it's a SecretParam
+ * - a literal value of the same type already defined in one of the .env files with key == param name
+ * - the value returned by interactively prompting the user
+ *   - the default value of the prompt comes from the SDK via param.default, which may be a literal value or a CEL expression
+ *   - if the default CEL expression is not resolvable--it depends on a param whose value is not yet known--we throw an error
+ *   - yes, this means that the same set of params may or may not throw depending on the order the SDK provides them to us in
+ *   - after prompting, the resolved value of the param is written to the most specific .env file available
+ */
+export async function resolveParams(
+  params: Param[],
+  projectId: string,
+  userEnvs: Record<string, ParamValue>
+): Promise<Record<string, ParamValue>> {
+  const paramValues: Record<string, ParamValue> = {};
+
+  for (const param of params) {
+    if (userEnvs.hasOwnProperty(param.param)) {
+      if (canSatisfyParam(param, userEnvs[param.param])) {
+        paramValues[param.param] = userEnvs[param.param];
+        continue;
+      } else {
+        throw new FirebaseError(
+          "Parameter " +
+            param.param +
+            " resolved to value from dotenv files " +
+            userEnvs[param.param] +
+            " of wrong type"
+        );
+      }
+    }
+
+    let paramDefault: ParamValue | undefined = param.default;
+    if (paramDefault && isCEL(paramDefault)) {
+      paramDefault = resolveDefaultCEL(param.type || "string", paramDefault, paramValues);
+    }
+    if (paramDefault && !canSatisfyParam(param, paramDefault)) {
+      throw new FirebaseError(
+        "Parameter " + param.param + " has default value " + paramDefault + " of wrong type"
+      );
+    }
+    paramValues[param.param] = await promptParam(param, paramDefault);
+  }
+
+  return paramValues;
+}
+
+/**
+ * Returns the resolved value of a user-defined Functions parameter.
+ * Functions params are defined by the output of the Functions SDK, but their value is not set until deploy-time.
+ *
+ * For most param types, we check the contents of the dotenv files first for a matching key, then interactively prompt the user.
+ * When the CLI is running in non-interactive mode or with the --force argument, it is an error for a param to be undefined in dotenvs.
+ */
+async function promptParam(param: Param, resolvedDefault?: ParamValue): Promise<ParamValue> {
+  if (param.type === "string") {
+    return promptStringParam(param, resolvedDefault as string);
+  } else if (param.type === "int") {
+    return promptIntParam(param, resolvedDefault as number);
```

Comment:
I know I'm being pedantic here, but we should be using `resolvedDefault as number | undefined`. As it currently stands, casting to number/string (I'm pretty sure) is performing a null assertion. This means someone can accidentally make promptIntParam and promptStringParam require the resolvedDefault parameter and everything will still compile but blow up at runtime

Reply:
Done